File: python/백준 온라인 저지/2022.02.11/랜선 자르기.py
import sys
input = sys.stdin.readline

k, n = map(int, input().split())
ren_line = []

for _ in range(k):
    ren_line.append(int(input()))

left, right = 1, max(ren_line)

while left <= right:
    mid = (left + right) // 2
    cnt = 0
    for i in range(k):
        cnt += ren_line[i] // mid
    
    if cnt < n:
        right = mid - 1
    else:
        left = mid + 1

if left == 1:
    print(left)
else:
    print(left-1)

File: Kotlin/Algorithm/src/랜선 자르기.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun main(args: Array<String>) = with(BufferedReader(InputStreamReader(System.`in`))) {
    val (k, n) = readLine().split(" ").map { it.toLong() }
    val arrInt = LongArray(k.toInt()) {0}

    for (i in 0 until k) arrInt[i.toInt()] = readLine().toLong()

    var left: Long = 1
    var right = arrInt.maxOrNull()!!
    var mid: Long
    var count: Long

    while (left <= right) {

        mid = (left + right) / 2
        count = 0

        for (i in 0 until k) count += (arrInt[i.toInt()] / mid)

        when {
            count < n -> {
                right = mid - 1
            }
            else -> {
                left = mid + 1
            }
        }
    }

    when (left) {
        1.toLong() -> println("$left")
        else -> println("${left - 1}")
    }

    if (left == 1.toLong()) {
        println("$left")
    } else {
        println("${left - 1}")
    }


}